### PR TITLE
topshell 0.7.7 (new formula)

### PR DESCRIPTION
**Pull request by bensleveritt**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

Adds TopShell 0.7.7, specifically built for macOS. First PR for brew, so apologies if I've missed anything.

### CODE REVIEW

File: Formula/topshell.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
+require "language/node"
+
+class Topshell < Formula
+  desc "Purely functional, reactive scripting language"
+  homepage "http://show.ahnfelt.net/topshell/"
+  url "https://github.com/topshell-language/topshell/archive/v0.7.7.tar.gz"
+  sha256 "a24c759b231fbd5aaed3409ed5d12e98ae2987cc27672fef27964a21017dccd6"
+
+  depends_on "node@8" => :build
```
Comment (fxcoudert) [inline Formula/topshell.rb]: We prefer not to accept new formulas that do not build with the latest version of compilers and tools, as it typically indicates lack of long-term support. In this case, `node@8` depends on Python 2, which will become unsupported on 2019-12-31, and will be removed from Homebrew. Can you migrate to latest `node`?

**Review by fxcoudert — CHANGES_REQUESTED** [review]

**Comment by bensleveritt** [comment]
Updated to remove the pinned version, and will use the latest, as requested.

**Comment by SMillerDev** [comment]
It doesn't require node to run anything?

**Comment by bensleveritt** [comment]
It uses Node to run the scripts, but it's packaged up with [pkg](https://www.npmjs.com/package/pkg), so it's a build dependency, rather than a runtime one.

**Comment by SMillerDev** [comment]
so, it ships it's own version of node?

**Comment by bensleveritt** [comment]
Yes, as I understand it, it bundles a node vm at during build, and will use that subsequently.

**Comment by SMillerDev** [comment]
@Homebrew/core bundling it's own version of node seems weird. Do we have any rules on this?

**Comment by fxcoudert** [comment]
> bundling it's own version of node

That's a strong NO. There must be a way to build with external node.

**Comment by chenrui333** [comment]
> That's a strong NO. There must be a way to build with external node.

+ 1

**Comment by SMillerDev** [comment]
@bensleveritt is there a way to build without bundling node?

**Review by zbeekman — CHANGES_REQUESTED** [review]
> so, it ships it's own version of node?

@Homebrew/core @bensleveritt @fxcoudert @SMillerDev:

I don't think this is an accurate description of what is going on. It's using [pkg](https://www.npmjs.com/package/pkg) to create an executable with a self contained runtime from Homebrew's node. From looking at the project website, the reported size still seems reasonable (10s of MB).

If this is still undesirable, it may be possible to install it via npm without packaging it.

File: Formula/topshell.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
+require "language/node"
+
+class Topshell < Formula
+  desc "Purely functional, reactive scripting language"
+  homepage "http://show.ahnfelt.net/topshell/"
```
[the remark this answers is not in the document]
Comment (bensleveritt) [reply Formula/topshell.rb]: Good point, fixed up

**Comment by bensleveritt** [comment]
@zbeekman Yes, that's the situation. It creates a runtime bundle from Homebrew node, in much the same way an Electron-based app would. This could put it in the cask category of brews, but as it's command-line service that can be built from source, it seemed to fit with homebrew-core.

Happy to close the PR if this doesn't fit.

**Comment by bensleveritt** [comment]
As I understand it, Topshell can't be installed via NPM, as it and the scripts are actually compiled with Scala. The Node VM is used to run the interpreted Scala.

https://github.com/topshell-language/topshell/wiki/FAQ

Perhaps @Ahnfelt can explain it better...

**Comment by iMichka** [comment]
This work has started in august, and has not made much progress since then. We have tens of requests for new formulae per week, and we can't invest so much time in fixing each build failure: if the package does not behave nicely from the beginning, it feels it will be a lot of work to make it work, and also a lot of work to maintain in the future.

I'll close this for the time being, until a working version is submitted. You can of course still maintain the formula in a separate tap if needed.

**Comment by bensleveritt** [comment]
That's fair. I've no idea what more I needed to do to make this work, but as there's heavy sketicism for inclusion

**Pull request closed without merging**